Ordering of descendant segment with multiple selectors (#269)
ietf-wg-jsonpath/draft-ietf-jsonpath-base: draft-ietf-jsonpath-base.md

A descendant segment produces zero or more descendants of the input value.

A descendant selector visits the input value and each of
its descendants such that:

* nodes of any array are visited in array order, and
* nodes are visited before their descendants.

The order in which the children of an object are visited is not stipulated, since
JSON objects are unordered.

Suppose the nodes, in the order visited, are `D1`, ..., `Dn` (where `n >= 1`).
Note that `D1` is the input value.

> For each `i` such that `1 <= i <= n`, the nodelist `Ri` is defined to be a result of applying
> the child segment `[<selectors>]` (with all its selectors) to the node `Di`.

Reviewer: ```suggestion
For each descendant node `Di` such that `1 <= i <= n`, the resulting nodelist `Ri` is defined to be a result of applying
the all of the selectors in the child segment `[<selectors>]` to the node `Di`.
```

> Suppose the nodes, in the order visited, are `D1`, ..., `Dn` (where `n >= 1`).
> Note that `D1` is the input value.
> 
> For each `i` such that `1 <= i <= n`, the nodelist `Ri` is defined to be a result of applying
> the child segment `[<selectors>]` (with all its selectors) to the node `Di`.
> Note that `Ri` is the concatenation of the nodelists from each of the selectors.
> 
> The result of the descendant selector is the concatenation of `R1`, ..., `Rn` (in that order).
> Note that only a single traversal of the input value and its descendants is required.

Reviewer: ```suggestion
Suppose the nodes, in the order visited, are `D1`, ..., `Dn` where `n >= 1`.
(Note that `D1` is the input value.)

For each node `Di` such that `1 <= i <= n`, the resulting nodelist `Ri` is defined to be a result of applying
the all of the selectors in the child segment `[<selectors>]` to the node `Di`.

The final result of the descendant selector is the concatenation of `R1`, ..., `Rn` (in that order).
```

I fear that

> Note that only a single traversal of the input value and its descendants is required.

may be getting too close to specifying implementation logic.

Author: Ok, I'll delete the single traversal sentence.

I didn't adopt the proposal because it talks about applying all the selectors in the child segment, but we need child segment semantics (specifically, concatenating the selector nodelists) instead of just selector semantics.

Reviewer: My suggestion also provides a more precise wording and adjusts the parenthetical phrases.

> it talks about applying all the selectors in the child segment, but we need child segment semantics (specifically, concatenating the selector nodelists) instead of just selector semantics

Then I suggest we remove mention of the selectors completely.

```
Suppose the nodes, in the order visited, are `D1`, ..., `Dn` where `n >= 1`.
(Note that `D1` is the input value.)

For each node `Di` such that `1 <= i <= n`, the intermediate nodelist `Ri` is the result of applying
the child segment `[<selectors>]` to the node `Di`.

The result of the descendant segment is the concatenation of the intermediate nodelists `R1`, ..., `Rn` (in that order).
```

We also need to be sure that in the "child segment" section, we define how the selector results are concatenated, which it seems we do:

> The resulting nodelist of a child segment is the concatenation of the nodelists from each of its selectors in the order that the selectors appear in the list.

Author: > Then I suggest we remove mention of the selectors completely.

I can live with that. See 3516bee163b81457e64be5dabe2d417893d38bbb.

Reviewer: Why are you not taking the rest of the suggestion?

Author: Ah, sorry, I should also have deleted "Note that `Ri` is the concatenation of the nodelists from each of the selectors.". Fixed in 0d25c700fb7e02ec7db97bf3acc3016085384fe5.

As for the rest of the suggestion, I prefer the current wording in the PR:
* I prefer to define `Ri` via universal quantification over `i`, rather than over `Di`.
* I don't see any advantage in introducing the extra term "intermediate nodelist".

Are you comfortable with approving now?

Author: Thanks for explaining your opinions. I beg to differ. Is this a blocker for you or can you now approve the PR?

| `$..j`   | `4` <br> `1` | `$['a'][2][0]['j']` <br> `$['o']['j']` | Alternative result |
| `$..[0]` | `5` <br> `{"j": 4}` | `$['a'][0]` <br> `$['a'][2][0]` | Array values       |
| `$..[0]` | `{"j": 4}` <br> `5` | `$['a'][2][0]` <br> `$['a'][0]` | Alternative result |

> | `$..[*]` <br> `$..*` | `{"j": 1, "k" : 2}` <br> `[5, 3, [{"j": 4}, {"k": 6}]]` <br> `1` <br> `2` <br> `5` <br> `3` <br> `[{"j": 4}, {"k": 6}]` <br> `{"j": 4}` <br> `4` <br> `{"k": 6}` <br> `6` | `$['o']` <br> `$['a']` <br> `$['o']['j']` <br> `$['o']['k']` <br> `$['a'][0]` <br> `$['a'][1]` <br> `$['a'][2]` <br> `$['a'][2][0]` <br> `$['a'][2][0]['j']` <br> `$['a'][2][1]` <br> `$['a'][2][1]['k']` | All values    |

Author: Perhaps this example is becoming a bit unwieldy and we should have more than one JSON document, each with their own table of queries. But, otherwise, I quite like keeping to a single JSON document and table per group of examples.

| `$..o`   | `{"j": 1, "k": 2}` | `$['o']` | Input value is visited |
| `$.o..[*, *]` | `1` <br> `2` <br> `2` <br> `1` | `$['o']['j']` <br> `$['o']['k']` <br> `$['o']['k']` <br> `$['o']['j']` | Non-deterministic ordering |

> | `$.a..[0, 1]`| `5` <br> `3` <br> `{"j": 4}` <br> `{"k": 6}` | `$['a'][0]` <br> `$['a'][1]` <br> `$['a'][0][2][0]` <br> `$['a'][0][2][1]` | Multiple segments |

Reviewer: Is it necessary to have the `.a` in here?  I thought we were trying to illustrate the multi-selector behavior.  It seems having the additional segment muddies the example.

Author: Yes, but it's only an artefact of having a single JSON document and group of examples per section of examples. See https://github.com/ietf-wg-jsonpath/draft-ietf-jsonpath-base/pull/269/files#r998358723. Let's see how this reads later.
